tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -139,8 +153,335 @@ impl ServerKey {
             }
         };
 
-        self.unchecked_add_assign(lhs, rhs);
-        self.full_propagate_parallelized(lhs);
+        if self.is_eligible_for_parallel_carryless_add(lhs.blocks.len()) {
+            self.unchecked_add_assign_parallelized_low_latency(lhs, rhs, AddExtraOne::No);
+        } else {
+            self.unchecked_add_assign(lhs, rhs);
+            self.full_propagate_parallelized(lhs);
+        }
+    }
+
+    pub fn add_parallelized_work_efficient<PBSOrder: PBSOrderMarker>(
+        &self,
+        ct_left: &RadixCiphertext<PBSOrder>,
+        ct_right: &RadixCiphertext<PBSOrder>,
+    ) -> RadixCiphertext<PBSOrder> {
+        let mut ct_res = ct_left.clone();
+        self.add_assign_parallelized_work_efficient(&mut ct_res, ct_right);
+        ct_res
+    }
+
+    pub fn add_assign_parallelized_work_efficient<PBSOrder: PBSOrderMarker>(
+        &self,
+        ct_left: &mut RadixCiphertext<PBSOrder>,
+        ct_right: &RadixCiphertext<PBSOrder>,
+    ) {
+        let mut tmp_rhs: RadixCiphertext<PBSOrder>;
+
+        let (lhs, rhs) = match (
+            ct_left.block_carries_are_empty(),
+            ct_right.block_carries_are_empty(),
+        ) {
+            (true, true) => (ct_left, ct_right),
+            (true, false) => {
+                tmp_rhs = ct_right.clone();
+                self.full_propagate_parallelized(&mut tmp_rhs);
+                (ct_left, &tmp_rhs)
+            }
+            (false, true) => {
+                self.full_propagate_parallelized(ct_left);
+                (ct_left, ct_right)
+            }
+            (false, false) => {
+                tmp_rhs = ct_right.clone();
+                rayon::join(
+                    || self.full_propagate_parallelized(ct_left),
+                    || self.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (ct_left, &tmp_rhs)
+            }
+        };
+
+        self.unchecked_add_assign_parallelized_low_latency(lhs, rhs, AddExtraOne::No);
```

Comment:
I seemed to understand work efficient does not mean the parallelized low latency ? so here we have the work efficient version calling the non efficient version if I'm reading this right ?

Reply:
yeah that is a typo :sweat_smile: :hot_face:

Reply:
fixed